Rename real to scalar_t.

ezyang opened the pull request:
This is necessary to allow us to use the complex header
which defines real (and is very sad if real is macro'ed).

We should also fix accreal, ureal, Real and REAL, but
only 'real' is the real blocker.

```
codemod -d aten/src/TH --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THC --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THNN --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THCUNN --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Rename real to scalar_t.

This is necessary to allow us to use the complex header
which defines real (and is very sad if real is macro'ed).

We should also fix accreal, ureal, Real and REAL, but
only 'real' is the real blocker.

```
codemod -d aten/src/TH --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THC --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THNN --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d aten/src/THCUNN --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
codemod -d torch/csrc/generic --extensions c,cc,cpp,cu,cuh,h,TARGETS,py,hpp '\breal\b' scalar_t
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fix HIPify as well.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

SsnL commented on aten/src/TH/generic/THTensorApply.hpp:
```diff
@@ -110,12 +110,12 @@
 //         TENSOR3 is index
 // Tests:
 //   1. index->size(d) <= src->size(d) for all d
-//   2. index->size(d) <= real->size(d) for all d != dim
```
LOL `codemod` isn't smart enough. Can you revert the changes in this comment?

SsnL commented on aten/src/TH/generic/THTensorLapack.cpp:
```diff
@@ -702,7 +702,7 @@ void THTensor_(potri)(THTensor *ra_, THTensor *a, const char *uplo)
 }
 
 /*
- Computes the Cholesky factorization with complete pivoting of a real symmetric
```
revert here too. is there an option in `codemod` that makes it ignore comments?

SsnL replied:
I checked `codemod` doc and there doesn't seem to be such an option

SsnL commented on aten/src/TH/generic/THTensorLapack.cpp:
```diff
@@ -820,17 +820,17 @@ void THTensor_(geqrf)(THTensor *ra_, THTensor *rtau_, THTensor *a)
                    &wkopt, -1, &info);
 
-  /* Allocate the workspace and call LAPACK to do the real work. */
+  /* Allocate the workspace and call LAPACK to do the scalar_t work. */
```
here too

SsnL commented on aten/src/TH/generic/THTensorLapack.cpp:
```diff
@@ -872,17 +872,17 @@ void THTensor_(orgqr)(THTensor *ra_, THTensor *a, THTensor *tau)
                    &wkopt, -1, &info);
 
-  /* Allocate the workspace and call LAPACK to do the real work. */
+  /* Allocate the workspace and call LAPACK to do the scalar_t work. */
```
here too

SsnL commented on aten/src/THC/THCReduce.cuh:
```diff
@@ -359,7 +359,7 @@ inline dim3 getContigReduceBlock(ptrdiff_t numSlices, int64_t reductionSize) {
   // is high, then we should increase block size for greater parallelism.
   // Aim for at least 32 warps per SM (assume 15 SMs; don't bother
-  // inquiring the real number for now).
+  // inquiring the scalar_t number for now).
```
here too: the real number

SsnL commented on aten/src/THC/generic/THCTensorSort.cu:
```diff
@@ -254,7 +254,7 @@ void THCTensor_(sortViaThrust)(THCState* state,
     SliceComp(sliceSize));
 
-  // Translate the global integer 0-based index to a per-slice real
+  // Translate the global integer 0-based index to a per-slice scalar_t
```
here too

SsnL commented on aten/src/THNN/generic/Col2Im.c:
```diff
@@ -28,7 +28,7 @@
 // out-of-bounds if you're computing backwards for a 1x1 input.
 //
-// "Now Edward," you might say, "the real problem is that you set output_padding
+// "Now Edward," you might say, "the scalar_t problem is that you set output_padding
```
the real problem :P

ssnl reviewed (commented):
Changes mostly looking good, except for a couple places where `codemod` shouldn't change things (e.g., in some comments). I commented at places I noticed, but I'm not sure if I didn't miss something. 

A <del>real</del> scalar_t concern I have is `THCNumerics.cuh` where we define 
```cpp
template <typename scalar_t>
static inline __host__ __device__ scalar_t powi(scalar_t a, scalar_t b);
```

gcc7 complains when I do 
```cpp
<source>:7:23: error: expected nested-name-specifier before 'double'

     template<typename double>

                       ^~~~~~
```

It seems that it doesn't cause problem currently. But I wonder if this will be problematic if in future someone include this header in some file.

SsnL commented:
Expanding all files gave me the most laggy chrome behavior I've ever experienced.

ezyang committed:
Remove scalar_t from template parameters; make them T instead.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fix incorrect translations of comments

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fix some multiline comment issues

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Here's how I'm addressing the comments:

* I grepped for `ag "typename scalar_t"` in the TH directories, and manually fixed all occurrences that popped up by replacing `scalar_t` with `T`
* Fixed the referenced comment issues
* Searched for "//.+scalar_t" to see if I missed any inline comment issues
* Grepped the diff for /*, which actually did bring up some more comment issues

Thinking about the template parameter problem, I'm actually not sure how wise it is to rename this as `scalar_t`, because that's our canonical name for our templated methods, which means that if we include any of the `template<typename scalar_t>` headers from inside generic TH, we'll twing the error. I guess this shouldn't actually happen.

ezyang committed:
Fix a missing T transform

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Needs fbcode changes.

SsnL commented:
New commits looks good! Let me know if you have the fbcode changes ready and I'll review.
